pkg/controller/iam/serviceaccountkey.go

```diff
@@ -0,0 +1,336 @@
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	}
+
+	req := s.serviceAccountKeyClient.Create(saPath, &iamv1.CreateServiceAccountKeyRequest{
+		KeyAlgorithm:   gcp.StringValue(cr.Spec.ForProvider.KeyAlgorithm),
+		PrivateKeyType: gcp.StringValue(cr.Spec.ForProvider.PrivateKeyType),
+	})
+
+	if fromProvider, err := req.Context(ctx).Do(); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	} else if err := populateSaKey(cr, fromProvider); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	} else if connDetails, err := getConnectionDetails(cr.Spec.ForProvider.PublicKeyType, fromProvider); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	} else if keyID, err := parseKeyIDFromRrn(fromProvider.Name); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateServiceAccountKey)
+	} else {
+		meta.SetExternalName(cr, keyID) // set external name to key id parsing it from Google Cloud API relative resource name
+
+		return managed.ExternalCreation{
+			ExternalNameAssigned: true,
+			ConnectionDetails:    connDetails,
+		}, nil
+	}
+}
+
+func (s *serviceAccountKeyExternalClient) Update(_ context.Context,
+	_ resource.Managed) (managed.ExternalUpdate, error) {
+	// ServiceAccountKeys are immutable, i.e.,GCP IAM Rest API does not provide an update method:
+	// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts.keys
+	return managed.ExternalUpdate{}, nil
+}
+
+func (s *serviceAccountKeyExternalClient) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.ServiceAccountKey)
+
+	if !ok {
+		return errors.New(errNotServiceAccountKey)
+	}
+
+	rrn, extNameExists, err := s.rrn.resourcePath(cr)
+
+	if err != nil {
+		return errors.Wrap(err, errDeleteServiceAccountKey)
+	}
+
+	if !extNameExists {
+		return nil
+	}
+
+	_, err = s.serviceAccountKeyClient.Delete(rrn).Context(ctx).Do()
+
+	if gcp.IsErrorNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, errDeleteServiceAccountKey)
+}
+
+// NewRelativeResourceNamer makes an instance of the RelativeResourceNamer
+// which is the only type that is allowed to know how to construct GCP resource names
+// for the IAM type.
+func newSaKeyRelativeResourceNamer(projectID string) saKeyrelativeResourceNamer {
+	return saKeyrelativeResourceNamer{
+		saNamer: RelativeResourceNamer{
+			projectName: projectID,
+		},
+	}
+}
+
+// saKeyrelativeResourceNamer allows the controller to generate the Google Cloud API relative resource name
+// for the service account key based on the external-name annotation.
+// https://cloud.google.com/apis/design/resource_names#relative_resource_name
+// The relative resource names for service account keys have the following format:
+// projects/{project_id}/serviceAccounts/{account}/keys/{key}
+type saKeyrelativeResourceNamer struct {
```

Comment:
A similar struct exists in `ServiceAccount` as well but I believe that one had some exceptions around naming. For `ServiceAccountKey`, the path to use is always `<sa-name>/keys/<sa-key-name>` so I think using directly `fmt.Sprintf(fmtKeyRelativeResourceName, saPath, extName)` whenever needed would be sufficient.